[v0.38] Feat / Add Connector for CoinZoom

TheHolyRoger opened the pull request:
**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Adds a connector for CoinZoom!

PtrckM commented:
- connector status is [green](https://github.com/CoinAlpha/hummingbot/pull/3112/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R36), new connectors should be yellow
- during creation of orders, bot unable to deduct and capture the right balance this only happens when you first start the bot.
![image](https://user-images.githubusercontent.com/54516858/112625124-73c60680-8e69-11eb-8f71-b532a78d2de4.png)
- unable to get the latest trade on the market (run time 40 minutes)
![image](https://user-images.githubusercontent.com/54516858/112625446-e33bf600-8e69-11eb-890f-0b8faf74c826.png)
- unable to display USD value 
![image](https://user-images.githubusercontent.com/54516858/112626423-23e83f00-8e6b-11eb-8531-7a12e532e9d1.png)

- unit testing
   - test_coinzoom_auth.py 0/2 = fail
   - test_coinzoom_exchange.py 0/1 = fail

TheHolyRoger commented:
@PtrckM regarding the balance updates, CoinZoom doesn't support any websocket balance updates, so we have to wait for polling.

Will get onto and fix the rest of your notes, thank you for testing!

TheHolyRoger commented:
@PtrckM my ticker "Last Trade" is updating, is yours not?

The text on that `ticker` is misleading, in all connectors I've seen (just checked coinbase, binance, blocktane and others) - it reports the last ticker price as reported by the exchange - which is what I'm seeing here, it matches the data from the URL:
https://api.stage.coinzoom.com/api/v1/public/marketwatch/ticker
---------------
Regarding the unit tests, are you setting your environment keys like so?
```bash
export COINZOOM_API_KEY=your_api_key
export COINZOOM_SECRET_KEY=your_secret_key
export COINZOOM_USERNAME=your_username
```
---------------
Can confirm I'm seeing empty USD balance value, investigating that now

TheHolyRoger commented:
@Nullably I've submitted a separate PR for the 0 value issue with the rate oracle here:
https://github.com/CoinAlpha/hummingbot/pull/3118

Rate Oracle also returns None price for any pairs that aren't on Binance or Coingecko e.g. the ZOOM token on CoinZoom, see screenshot:
<img width="285" alt="Screenshot 2021-03-26 at 14 57 57" src="https://user-images.githubusercontent.com/39387497/112650811-af4dda00-8e43-11eb-8828-a8333461f74a.png">

PtrckM commented:
- :white_check_mark: the oracle PR seems fix the USD value as long as the token exists on binance/coingecko 
- :white_check_mark: last_trade has delay but bot able to sync after sometime. 

![image](https://user-images.githubusercontent.com/54516858/112712563-6d886680-8f0b-11eb-874f-9e4f75687dc8.png)

![image](https://user-images.githubusercontent.com/54516858/112712905-55b1e200-8f0d-11eb-9b6c-abf8c26197ca.png)

PtrckM commented:
- unit testing `test_coinzoom_auth.py` = `test_ws_auth` this one is failing on me.
```
test_rest_auth (test.connector.exchange.coinzoom.test_coinzoom_auth.TestAuth) ... ok
test_ws_auth (test.connector.exchange.coinzoom.test_coinzoom_auth.TestAuth) ... ERROR

======================================================================
ERROR: test_ws_auth (test.connector.exchange.coinzoom.test_coinzoom_auth.TestAuth)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/coinzoom/test/connector/exchange/coinzoom/test_coinzoom_auth.py", line 53, in test_ws_auth
    subscribed = self.ev_loop.run_until_complete(self.ws_auth())
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/ubuntu/coinzoom/test/connector/exchange/coinzoom/test_coinzoom_auth.py", line 43, in ws_auth
    return True
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 55, in __aexit__
    self._do_exit(exc_type)
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 92, in _do_exit
    raise asyncio.TimeoutError
asyncio.exceptions.TimeoutError: 
```

[test_coinzoom_auth.log](https://github.com/CoinAlpha/hummingbot/files/6215583/test_coinzoom_auth.log)

TheHolyRoger commented:
> * unit testing `test_coinzoom_auth.py` = `test_ws_auth` this one is failing on me.
> 
> ```
> test_rest_auth (test.connector.exchange.coinzoom.test_coinzoom_auth.TestAuth) ... ok
> test_ws_auth (test.connector.exchange.coinzoom.test_coinzoom_auth.TestAuth) ... ERROR
> 
> ======================================================================
> ERROR: test_ws_auth (test.connector.exchange.coinzoom.test_coinzoom_auth.TestAuth)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/ubuntu/coinzoom/test/connector/exchange/coinzoom/test_coinzoom_auth.py", line 53, in test_ws_auth
>     subscribed = self.ev_loop.run_until_complete(self.ws_auth())
>   File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
>     return future.result()
>   File "/home/ubuntu/coinzoom/test/connector/exchange/coinzoom/test_coinzoom_auth.py", line 43, in ws_auth
>     return True
>   File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 55, in __aexit__
>     self._do_exit(exc_type)
>   File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 92, in _do_exit
>     raise asyncio.TimeoutError
> asyncio.exceptions.TimeoutError: 
> ```
> 
> [test_coinzoom_auth.log](https://github.com/CoinAlpha/hummingbot/files/6215583/test_coinzoom_auth.log)

@PtrckM Thanks and sorry! Forgot I changed the constant there, fixed now :)

rxlxrxsx commented:
Confirmed that all tests passed during unit testing.

TheHolyRoger commented on hummingbot/connector/exchange/coinzoom/coinzoom_constants.py:
```diff
@@ -0,0 +1,60 @@
+    https://api-markets.coinzoom.com/
+    """
+    EXCHANGE_NAME = "coinzoom"
+    # REST_URL = "https://api.coinzoom.com/api/v1/public"
```
Need to switch from production to stage before this PR is merged.

TheHolyRoger replied:
Done.

rxlxrxsx commented:
Order book price and volume on bot is not totally in sync with the exchange. Tested on BTC-USD and EOS-USD market

https://user-images.githubusercontent.com/64377055/113101205-07635280-922f-11eb-8585-d6a1b4bb796e.mp4

https://user-images.githubusercontent.com/64377055/113101239-0f22f700-922f-11eb-8d7e-6b96adfabdc7.mp4

Also for clarification, is it expected for Coinzoom to not have any exchange_trade_id on executed trades?
<img width="788" alt="Screen Shot 2021-03-31 at 2 00 48 PM" src="https://user-images.githubusercontent.com/64377055/113101495-5d37fa80-922f-11eb-8315-a499593f6357.png">

TheHolyRoger commented:
> Order book price and volume on bot is not totally in sync with the exchange. Tested on BTC-USD and EOS-USD market
> Also for clarification, is it expected for Coinzoom to not have any exchange_trade_id on executed trades?
@rxlxrxsx the sync issue is down to coinzoom's websocket implementation as far as i'm aware - don't think it's hummingbot.

Regarding exchange_trade_id not being logged in the CSV file - I'm getting the same results on **all** connectors on dev branch...?

TheHolyRoger commented:
@rxlxrxsx thank you fixed the exchange_trade_id now!

dennisocana commented:
@CrimsonJacket please do a code review for this connector. Thanks

RC-13 reviewed (dismissed):
Tested and confirmed working by QA, requesting a review from the engineering team.

TheHolyRoger commented:
@RC-13 Added the logo sorry it dismissed the review

CrimsonJacket commented on hummingbot/connector/exchange/coinzoom/coinzoom_auth.py:
```diff
@@ -0,0 +1,31 @@
+            "Content-Type": "application/json",
+            "Coinzoom-Api-Key": str(self.api_key),
+            "Coinzoom-Api-Secret": str(self.secret_key),
+            "User-Agent": f"hummingbot ZoomMe: {self.username}"
```
@dennisocana Could you confirm this HTTP header? I assume this is also for Hummingbot Broker verification.

TheHolyRoger replied:
@CrimsonJacket the User-Agent header? It's required for authentication

RC-13 reviewed (dismissed):
Re-approving PR

CrimsonJacket commented on hummingbot/connector/exchange/coinzoom/coinzoom_constants.py:
```diff
@@ -0,0 +1,60 @@
+    # Only used when nothing is received from WS
+    SHORT_POLL_INTERVAL = 5.0
+    # CoinZoom poll interval can't be too long since we don't get balances via websockets
+    LONG_POLL_INTERVAL = 8.0
```
Instead of using the `LONG_POLL_INTERVAL` to determine when the bot will update its balance, would it be better to simply just request for the balance every time there is an order update? 

@CoinAlpha/qa-team Were you able to capture user balance not updating?

TheHolyRoger replied:
@CrimsonJacket we could do, but since this would be a REST request should I implement cachetools or something so as not to spam requests?

RC-13 replied:
Yes, during the initial start-up, took around 6 sec for the available balance to update.

CrimsonJacket replied:
I do not think user data via the private websocket channel will be received that often. But I understand the concern with rate limits.

If it is a concern, maybe including a api_request throttler like in kraken would be good.

CrimsonJacket replied:
@RC-13 How about between order proposals? Say after orders have been cancelled? There should be an edge case where, orders get cancelled but the available balance does not get updated, resulting in a "insufficient" balance being reported.

RC-13 replied:
@CrimsonJacket Right, sorry I forgot about this. We don't have a warning but it does experience inconsistency with order creation due to delay with updating of balance (similar behavior on other connectors). It's likely to be seen when running multiple orders and balance is not sufficient to cover all the levels.

TheHolyRoger replied:
@CrimsonJacket I will make these changes later tonight!

TheHolyRoger commented:
@RC-13 @CrimsonJacket made these changes now if you could review again please?

PtrckM reviewed (approved):
Re-approving no issue found, tested and confirmed working by QA, requesting a review from the engineering team.

CrimsonJacket reviewed (approved):
LGTM 👍
